sources/api/corndog/src/main.rs

```diff
@@ -104,23 +109,104 @@ where
+    if !key.chars().all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' || c == '*') {
+        return error::InvalidSysctlKeySnafu { key: key.to_string() }.fail();
+    }
+
+    // Ensure value doesn't contain newlines which could break the config file format
+    if value.contains('\n') {
+        return error::InvalidSysctlValueSnafu { 
+            key: key.to_string(),
+            value: value.to_string() 
+        }.fail();
+    }
+
+    Ok(())
+}
+
+/// Creates a backup of the existing sysctl config file
+fn backup_sysctl_config() -> Result<Option<NamedTempFile>> {
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+    
+    if !Path::new(SYSCTL_CONFIG).exists() {
+        return Ok(None);
+    }
+
+    let mut temp_file = NamedTempFile::new().context(error::TempFileSnafu)?;
+    let contents = fs::read_to_string(SYSCTL_CONFIG).context(error::ReadSysctlConfigSnafu)?;
+    temp_file.write_all(contents.as_bytes()).context(error::WriteTempFileSnafu)?;
+    
+    Ok(Some(temp_file))
+}
+
+/// Restores sysctl config from backup
+fn restore_sysctl_config(backup: NamedTempFile) -> Result<()> {
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+    
+    fs::copy(backup.path(), SYSCTL_CONFIG).context(error::RestoreBackupSnafu)?;
+    Ok(())
+}
+
+/// Restarts systemd-sysctl service
+fn restart_sysctl() -> Result<()> {
+    let status = Command::new("systemctl")
+        .args(["restart", "systemd-sysctl"])
+        .status()
+        .context(error::SystemctlSnafu)?;
+
+    if !status.success() {
+        return error::SystemctlFailedSnafu.fail();
+    }
+    
+    Ok(())
+}
+
+/// Applies the requested sysctls to the system using systemd-sysctl.
+fn set_sysctls<K>(sysctls: HashMap<K, String>) -> Result<()>
+where
+    K: AsRef<str>,
+{
+    // TODO: make the priority configurable?
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+
+    // Validate all entries first
+    for (key, value) in &sysctls {
+        validate_sysctl_entry(key, value)?;
+    }
+
+    // Create backup of existing config
+    let backup = backup_sysctl_config()?;
+
+    // Write new config
+    let mut config = String::new();
     for (key, value) in sysctls {
-        let key = key.as_ref();
-        let path = sysctl_path(key);
-        if let Err(e) = fs::write(path, value) {
-            // We don't fail because sysctl keys can vary between kernel versions and depend on
-            // loaded modules.  It wouldn't be possible to deploy settings to a mixed-kernel fleet
-            // if newer sysctl values failed on your older kernels, for example, and we believe
-            // it's too cumbersome to have to specify in settings which keys are allowed to fail.
-            error!("Failed to write sysctl value '{}': {}", key, e);
+        config.push_str(&format!("{}={}\n", key.as_ref(), value));
+    }
+    
+    fs::write(SYSCTL_CONFIG, config).context(error::WriteSysctlConfigSnafu)?;
```

Comment:
I need to do some double-checking around:

- Does this actually append to the file as written, or does it clobber it?
- If we append and we write duplicate sysctls to the same config file, how does systemd-sysctl handle that? (Does the most recent sysctl win?)

I haven't yet been able to get a local build working (turns out my hard drive is smaller than I thought!) but I'll investigate these shortly.

Reply:
Answers:

- This clobbers the file (must fix)
- systemd-sysctl will use the last value in a file for the same sysctl. For example, a config file with the contents:

```
dev.raid.speed_limit_max = 199999 
dev.raid.speed_limit_max = 199998 
garbage                           
dev.raid.speed_limit_max = 199997
```

and running `systemctl restart systemd-sysctl` will set `dev.raid.speed_limit_max` to `199997` and report an error when attempting to parse the "garbage" line - the return code of the systemd unit will be 1 but all of the correctly specified sysctl k/v pairs will have been applied.

systemd-sysctl also will deduplicate multiple lines and only write when there are changes to be made, so in theory endlessly appending to a file _should_ be a reasonable way to do this.

Reply:
Output from running systemd-sysctl with the above file in support of the above conclusions:

```
bash-5.1# SYSTEMD_LOG_LEVEL=debug /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/systemd-sysctl
Parsing /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/sysctl.d/50-default.conf
Parsing /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/sysctl.d/50-pid-max.conf
Parsing /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/sysctl.d/80-release.conf
Parsing /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/sysctl.d/90-kubelet.conf
Parsing /etc/sysctl.d/90-primary_interface.conf
Parsing /etc/sysctl.d/99-zzz-override_cilium.conf
Overwriting earlier assignment of net/ipv4/conf/all/rp_filter at '/etc/sysctl.d/99-zzz-override_cilium.conf:7'.
Parsing /etc/sysctl.d/99-zzz-zzz-testing.conf
Overwriting earlier assignment of dev/raid/speed_limit_max at '/etc/sysctl.d/99-zzz-zzz-testing.conf:2'.
/etc/sysctl.d/99-zzz-zzz-testing.conf:3: Line is not an assignment, ignoring: garbage
Overwriting earlier assignment of dev/raid/speed_limit_max at '/etc/sysctl.d/99-zzz-zzz-testing.conf:4'.
...
Setting '/proc/sys/dev/raid/speed_limit_max' to '199997'
No change in value '199997', suppressing write
```